httpcore5/src/main/java/org/apache/hc/core5/http/io/ResponseOutOfOrderStrategy.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * ====================================================================
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+
+package org.apache.hc.core5.http.io;
+
+import org.apache.hc.core5.http.ClassicHttpRequest;
+
+/**
+ * Represents a strategy to determine how frequently the client should check for an out of order response.
+ * An out of order response is sent before the server has read the full request, which may result in the
+ * request becoming deadlocked and eventually timing out.
+ *
+ * @since 5.1
+ */
+public interface ResponseOutOfOrderStrategy {
+
+    /**
+     * Returns the interval in bytes describing how often the socket is checked for an out-of-order response.
+     * A value of zero results in checks being disabled, this may result in requests blocking until the socket
+     * timeout is reached while a server sends an out-of-order response.
+     */
+    int determineResponseCheckIntervalBytes(ClassicHttpRequest request);
```

Comment:
Opposed to the JIRA ticket, the Socket is gone. Why?

Reply:
Based on HTTPCORE-640 I don't think it's helpful to check if the socket is an instance of SSLSocket anymore.
The socket object could be helpful if we want to check the send and receive buffer sizes when they are explicitly set, but I'm not sure exactly how that would work, and in most cases the system defaults are used.

Reply:
@ok2c WDYT, do we need the socket here?